cmake/functions.cmake

```diff
@@ -28,23 +28,55 @@ endif()
 # that is linked into the Windows executables
 #
 function(ml_generate_resources _target)
-  if(NOT WIN32)
-    return()
+
+  set(ML_VERSION_STR "${ML_VERSION_NUM}")
+
+  if(ENV{VERSION_QUALIFIER})
+    set(ML_VERSION_STR "${ML_VERSION_STR}-$ENV{VERSION_QUALIFIER}")
+  endif()
+
+  if(NOT ENV{SNAPSHOT} STREQUAL no)
+    set(ML_VERSION_STR "${ML_VERSION_STR}-SNAPSHOT")
+  endif()
+
+  if(${ML_VERSION_STR} MATCHES "([0-9.]+)")
+    set(ML_VERSION ${CMAKE_MATCH_1})
+  endif()
+  string(REPLACE "." "," ML_VERSION "${ML_VERSION}")
+
+  set(ML_PATCH "0")
+
+  execute_process(COMMAND git -c core.fileMode=false update-index -q --refresh ERROR_FILE /dev/null OUTPUT_FILE /dev/null)
+  execute_process(COMMAND git -c core.fileMode=false diff-index --quiet HEAD --  RESULT_VARIABLE UNCOMMITTED_CHANGES)
+
+  if(UNCOMMITED_CHANGES EQUAL 0)
+    set(ML_FILEFLAGS "0")
+  else()
+    set(ML_FILEFLAGS "VS_FF_PRIVATEBUILD")
   endif()
-  set( ${_target}_LINKFLAGS ${CMAKE_CURRENT_BINARY_DIR}/${_target}.res )
-  set_target_properties( ${_target} PROPERTIES LINK_FLAGS ${${_target}_LINKFLAGS} )
-  execute_process(COMMAND bash -c "${CMAKE_SOURCE_DIR}/mk/make_rc_defines.sh ${_target}.exe" OUTPUT_VARIABLE
-    RC_DEFINES OUTPUT_STRIP_TRAILING_WHITESPACE)
-  file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/tmp.sh "rc -nologo ${CPPFLAGS} ${RC_DEFINES} -Fo${_target}.res ${CMAKE_SOURCE_DIR}/mk/ml.rc")
-  add_custom_target(
-    ${_target}.res
-    DEPENDS ${CMAKE_SOURCE_DIR}/mk/ml.rc ${CMAKE_SOURCE_DIR}/gradle.properties ${CMAKE_SOURCE_DIR}/mk/ml.ico ${CMAKE_SOURCE_DIR}/mk/make_rc_defines.sh ${CMAKE_CURRENT_BINARY_DIR}/tmp.sh
-    COMMAND bash -c ${CMAKE_CURRENT_BINARY_DIR}/tmp.sh
-    )
-  add_dependencies(${_target} ${_target}.res)
 
-  set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES ${CMAKE_CURRENT_BINARY_DIR}/tmp.sh)
-  set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES ${CMAKE_CURRENT_BINARY_DIR}/${_target}.res)
+  if("${_target}" MATCHES ".dll$")
+    set(ML_FILETYPE "VFT_DLL")
+  else()
+    set(ML_FILETYPE "VFT_APP")
+  endif()
+
+  set(ML_FILENAME ${_target})
+
+  if(${_target} MATCHES "([^.]+).")
+    set(ML_NAME ${CMAKE_MATCH_1})
+  endif()
+  set(ML_YEAR ${BUILD_YEAR})
+  set(ML_ICON ${CMAKE_SOURCE_DIR}/mk/ml.ico)
+
+  configure_file(
+	"${CMAKE_SOURCE_DIR}/mk/ml.rc.in"
+	"${CMAKE_CURRENT_BINARY_DIR}/${ML_NAME}.rc"
+        @ONLY
+        )
+
+  set(ML_FILEFLAGS ${ML_FILEFLAGS} PARENT_SCOPE)
```

Comment:
This is OK, but it does now mean that in addition to the resource file all our C++ code is getting compiled with this extra symbol defined.  I'm sure it won't be a problem as none of our source will contain this string.

It might be nicer to substitute it _only_ into the resource file using the `@ML_FILEFLAGS@` syntax.  Don't bother about doing this in this PR, because we want to get the overall feature branch merged soon.  But maybe you could have a look at tidying this up next week in a separate PR against the main branch.